components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/AuthUtil.java

```diff
@@ -0,0 +1,209 @@
+					//Remove from the first level token cache as well.
+					getGatewayTokenCache().remove(apiKey);
+				}
+				return info;
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * write to cache
+	 *
+	 * @param info
+	 * @param apiKey
+	 * @param cacheKey
+	 */
+	public static void putCache(APIKeyValidationInfoDTO info, String apiKey, String cacheKey) {
+
+		//Get the tenant domain of the API that is being invoked.
+		String tenantDomain =
+				PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantDomain();
+
+		//Add to first level Token Cache.
+		getGatewayTokenCache().put(apiKey, tenantDomain);
+		//Add to Key Cache.
+		getGatewayKeyCache().put(cacheKey, info);
+
+		//If this is NOT a super-tenant API that is being invoked
+		if (!MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+			//Add the tenant domain as a reference to the super tenant cache so we know from which tenant cache
+			//to remove the entry when the need occurs to clear this particular cache entry.
+			try {
+				PrivilegedCarbonContext.startTenantFlow();
+				PrivilegedCarbonContext.getThreadLocalCarbonContext().
+						setTenantDomain(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME, true);
+
+				getGatewayTokenCache().put(apiKey, tenantDomain);
+			} finally {
+				PrivilegedCarbonContext.endTenantFlow();
+			}
+		}
+
+	}
+
+	protected static Cache getGatewayKeyCache() {
+		String apimGWCacheExpiry =
+				ServiceReferenceHolder.getInstance().getAPIManagerConfiguration().
+						getFirstProperty(APIConstants.TOKEN_CACHE_EXPIRY);
+		if (!isGatewayKeyCacheInitialized && apimGWCacheExpiry != null) {
+			isGatewayKeyCacheInitialized = true;
+			return Caching.getCacheManager(APIConstants.API_MANAGER_CACHE_MANAGER)
+			              .createCacheBuilder(APIConstants.GATEWAY_KEY_CACHE_NAME).
+					              setExpiry(CacheConfiguration.ExpiryType.MODIFIED,
+					                        new CacheConfiguration.Duration(TimeUnit.SECONDS,
+					                                                        Long.parseLong(
+							                                                        apimGWCacheExpiry)))
+			              .
+					              setExpiry(CacheConfiguration.ExpiryType.ACCESSED,
+					                        new CacheConfiguration.Duration(TimeUnit.SECONDS,
+					                                                        Long.parseLong(
+							                                                        apimGWCacheExpiry)))
+			              .setStoreByValue(false).build();
+		} else {
+			return Caching.getCacheManager(APIConstants.API_MANAGER_CACHE_MANAGER)
+			              .getCache(APIConstants.GATEWAY_KEY_CACHE_NAME);
+		}
+	}
+
+	protected static Cache getGatewayTokenCache() {
+		Cache cache = null;
+		try {
+			javax.cache.CacheManager manager =
+					Caching.getCacheManager(APIConstants.API_MANAGER_CACHE_MANAGER);
+			cache = manager.getCache(APIConstants.GATEWAY_TOKEN_CACHE_NAME);
+		} catch (Exception e) {
+			log.error(
```

Comment:
any reason for catching generic exceptions? Also when log take e.getMessage(), e otherwise stack trace may print twice